File: clients/typescript/src/model/debug-data.ts
import { Vec2Double } from "./vec2-double";
import { Stream } from "../stream";

/**
 * TODO - Document
 */
export abstract class DebugData {
    /**
     * Write DebugData to output stream
     */
    abstract writeTo(stream: Stream): Promise<void>;

    /**
     * Read DebugData from input stream
     */
    static async readFrom(stream: Stream): Promise<DebugData> {
        const tag = await stream.readInt();
        if (tag == DebugData.Circle.TAG) {
            return await DebugData.Circle.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}

export namespace DebugData {
    /**
     * TODO - Document
     */
    export class Circle extends DebugData {
        /**
         * TODO - Document
         */
        pos: Vec2Double
        /**
         * TODO - Document
         */
        radius: number
    
        constructor(pos: Vec2Double, radius: number) {
            super();
            this.pos = pos;
            this.radius = radius;
        }
    
        /**
         * Read Circle from input stream
         */
        static async readFrom(stream: Stream): Promise<DebugData.Circle> {
            let pos;
            pos = await Vec2Double.readFrom(stream);
            let radius;
            radius = await stream.readDouble();
            return new Circle(pos, radius)
        }
    
        /**
         * Write Circle to output stream
         */
        async writeTo(stream: Stream) {
            await stream.writeInt(Circle.TAG);
            let pos = this.pos;
            await pos.writeTo(stream);
            let radius = this.radius;
            await stream.writeDouble(radius);
        }
    }
    
    export namespace Circle {
        export const TAG = 0;
    }
}

File: clients/ruby/model/debug_data.rb
require './model/vec2_double'

module Model

# TODO - Document
class DebugData
    # Read DebugData from input stream
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == DebugData::Circle::TAG
            return DebugData::Circle.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    # TODO - Document
    class Circle
        TAG = 0
    
        # TODO - Document
        attr_accessor :pos
        # TODO - Document
        attr_accessor :radius
    
        def initialize(pos, radius)
            @pos = pos
            @radius = radius
        end
    
        # Read Circle from input stream
        def self.read_from(stream)
            pos = Model::Vec2Double.read_from(stream)
            radius = stream.read_double()
            Circle.new(pos, radius)
        end
    
        # Write Circle to output stream
        def write_to(stream)
            stream.write_int(TAG)
            @pos.write_to(stream)
            stream.write_double(@radius)
        end
    
        def to_s
            string_result = "Circle { "
            string_result += "pos: "
            string_result += @pos.to_s
            string_result += ", "
            string_result += "radius: "
            string_result += @radius.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
end

end